cirq-ft/cirq_ft/algos/phase_estimation_test.py

```diff
@@ -0,0 +1,114 @@
+# Copyright 2021 The Cirq Developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from attr import frozen
+import numpy as np
+from cirq_ft import infra
+from cirq._compat import cached_property
+
+import cirq
+import pytest
+
+from cirq_ft.algos import phase_estimation as pe
+
+
+@frozen
+class BasicPrep(infra.GateWithRegisters):
+    control_register: infra.Register
+
+    @cached_property
+    def registers(self) -> infra.Registers:
+        return infra.Registers([self.control_register])
+
+    def decompose_from_registers(
+        self, context: cirq.DecompositionContext, **quregs
+    ) -> cirq.OP_TREE:
+        yield cirq.H.on_each(quregs[self.control_register.name])
+
+
+@frozen
+class Lamda(infra.GateWithRegisters):
+    control_register: infra.Register
+    eigen_vector_register: infra.Register
+    theta: int
+
+    @cached_property
+    def registers(self) -> infra.Registers:
+        return infra.Registers([self.control_register, self.eigen_vector_register])
+
+    @cached_property
+    def control_register_name(self):
+        return self.control_register.name
+
+    @cached_property
+    def eigenvector_register_name(self):
+        return self.eigen_vector_register.name
+
+    def decompose_from_registers(
+        self, context: cirq.DecompositionContext, **quregs
+    ) -> cirq.OP_TREE:
+        control_register_qubits = quregs[self.control_register_name]
+        eigenvector_register_qubits = quregs[self.eigenvector_register_name]
+        U = cirq.Z ** (2 * self.theta)
+        bits_of_precision = len(control_register_qubits)
+        yield [
+            cirq.ControlledGate(U).on(bit, *eigenvector_register_qubits)
+            ** (2 ** (bits_of_precision - i - 1))
+            for i, bit in enumerate(control_register_qubits)
+        ]
+
+
+@frozen
+class KitaevPhaseEstimation(pe.PhaseEstimation):
+    control_bitsize: int
+    eigenvector_bitsize: int
+    theta: int
```

Comment:
How about changing the interface to something like:

```python
@frozen
class KitaevPhaseEstimation(cirq.GateWithRegisters):
    u: cirq.Gate
    precision: int

```

Also, move `KitaevPhaseEstimation` to `phase_estimation.py` and test it using multiple different unitaries `u` ?

Reply:
do we need the the number of bits of the eigenvector?

Reply:
I removed it as discussed today.